transports/webtransport-websys/src/connection.rs

```diff
@@ -0,0 +1,647 @@
+        .map_err(Error::from_js_value)?;
+
+        let val = parse_reader_response(&val)
+            .map_err(Error::from_js_value)?
+            .ok_or_else(|| Error::JsError("incoming_bidirectional_streams closed".to_string()))?;
+
+        let bidi_stream = to_js_type::<WebTransportBidirectionalStream>(val)?;
+        let stream = Stream::new(bidi_stream)?;
+
+        Poll::Ready(Ok(stream))
+    }
+
+    /// Closes the session.
+    ///
+    /// This closes the streams also and they will return an error
+    /// when they will be used.
+    fn close_session(&mut self) {
+        if !self.closed {
+            detach_promise(self.incoming_streams_reader.cancel());
+            self.session.close();
+            self.closed = true;
+        }
+    }
+}
+
+/// WebTransport native multiplexing
+impl StreamMuxer for Connection {
+    type Substream = Stream;
+    type Error = Error;
+
+    fn poll_inbound(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Result<Self::Substream, Self::Error>> {
+        self.poll_incoming_bidirectional_streams(cx)
+    }
+
+    fn poll_outbound(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Result<Self::Substream, Self::Error>> {
+        self.poll_create_bidirectional_stream(cx)
+    }
+
+    fn poll_close(
+        mut self: Pin<&mut Self>,
+        _cx: &mut Context<'_>,
+    ) -> Poll<Result<(), Self::Error>> {
+        self.close_session();
+        Poll::Ready(Ok(()))
+    }
+
+    fn poll(
+        self: Pin<&mut Self>,
+        _cx: &mut Context<'_>,
+    ) -> Poll<Result<StreamMuxerEvent, Self::Error>> {
+        Poll::Pending
+    }
+}
+
+impl Drop for Connection {
+    fn drop(&mut self) {
+        self.close_session();
+    }
+}
+
+impl ConnectionSend {
+    pub(crate) fn new(conn: Connection) -> Self {
+        ConnectionSend {
+            inner: SendWrapper::new(conn),
+        }
+    }
+}
+
+impl StreamMuxer for ConnectionSend {
+    type Substream = StreamSend;
+    type Error = Error;
+
+    fn poll_inbound(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Result<Self::Substream, Self::Error>> {
+        let stream = ready!(self.get_mut().inner.poll_inbound_unpin(cx))?;
+        Poll::Ready(Ok(StreamSend::new(stream)))
+    }
+
+    fn poll_outbound(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Result<Self::Substream, Self::Error>> {
+        let stream = ready!(self.get_mut().inner.poll_outbound_unpin(cx))?;
+        Poll::Ready(Ok(StreamSend::new(stream)))
+    }
+
+    fn poll_close(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.get_mut().inner.poll_close_unpin(cx)
+    }
+
+    fn poll(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Result<StreamMuxerEvent, Self::Error>> {
+        self.get_mut().inner.poll_unpin(cx)
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
I was running manually these tests when I was developing the crate to understand the behavior of the JS API. I find them very useful. We can remove them from the CI, but if it is ok I would like to keep them in the crate.

Reply:
If we remove them from CI, I am worried that they'll get stale. One of the next items on our roadmap is to implement a WebTransport server. Once we have that, perhaps we can re-introduce them together with something like https://docs.rs/headless_chrome/1.0.5/headless_chrome/?

Thoughts?

Reply:
In that case the tests will written in the WebTransport server crate? If yes, then we will not be able to test some lower level cases, for example `poll_outbound_error_after_connection_close` or `allow_read_after_closing_writer`.

Reply:
I have an idea which I do not really like, but it can play very well when WebTransport server is implemented.
Imagine these test cases within `libp2p-webtransport-websys`:

```rust
#[cfg(test)]
mod tests {
    // ...

    #[test]
    fn wasm_tests() {
        // This will start the echo ever and will be written with
        // the native libp2p-webtransport.
        spawn_echo_server();

        // This will run `wasm-pack test --chrome --headless`
        let exit_code = run_wasm_pack_test_chromium();
        assert_eq!(exit_code, 0);

        // This will run `wasm-pack test --firefox --headless`
        let exit_code = run_wasm_pack_test_firefox();
        assert_eq!(exit_code, 0);
    }

    #[wasm_bindgen_test]
    async fn single_conn_single_stream() {
        // ...
    }

    // Other wasm-bindgen tests
}
```

With the above, developers and the CI will only need to run `cargo test`.

Thoughts?

Reply:
Also, when the native libp2p-webtransport is ready, are they going to be merged? In this case should I rename this crate to `libp2p-webtransport` and do something like this in lib.rs?

```rust
#[cfg(target_arch = "wasm32")]
mod websys;

#[cfg(target_arch = "wasm32")]
pub use crate::websys::*;
```

Also, should the `Transport` be `Send` in this case? Now I hide the `Send` wrapper from the user and `boxed` will give it indirectly.

Reply:
@thomaseizinger This is what i meant: https://github.com/oblique/rust-libp2p/commit/1833d0c26734757d761fef8be383e346ede58286

and the job output looks like this: https://github.com/oblique/rust-libp2p/actions/runs/5296669698/jobs/9587930247

If you like it, I can add the commit it in this PR.

Reply:
@thomaseizinger And for making `Transport` `Send`, I meant this: https://github.com/oblique/rust-libp2p/commit/bb55ccab82c9410f0f06641a3d16a5a72d0c3d53

If you prefer it, I can add the commit in this PR.

Reply:
> Also, when the native libp2p-webtransport is ready, are they going to be merged? In this case should I rename this crate to `libp2p-webtransport` and do something like this in lib.rs?
> 
> ```rust
> #[cfg(target_arch = "wasm32")]
> mod websys;
> 
> #[cfg(target_arch = "wasm32")]
> pub use crate::websys::*;
> ```
> 
> Also, should the `Transport` be `Send` in this case? Now I hide the `Send` wrapper from the user and `boxed` will give it indirectly.

I think it is fine for these to be separate crates. They wouldn't really share any dependencies so I don't see why they should be the same crate.

Reply:
> In that case the tests will written in the WebTransport server crate? If yes, then we will not be able to test some lower level cases, for example `poll_outbound_error_after_connection_close` or `allow_read_after_closing_writer`.

Ideally, we only test the public API of such a crate. Unless I am missing something, the two tests you linked to **do** use only the public API (via the `StreamMuxer` interface) so I think that should be fine?

Reply:
In regards to the actual tests, I'd be okay with the following:

- Introduce a new directory `wasm-tests` at the top-level
- Put all sorts of code in there that is necessary to set up our wasm tests, for example, the go echo server whilst we don't have a native Rust implementation of webtransport
- These tests would also go in there, probably as a separate crate?
- Then we will probably need some sort of shell script to run all the tests in there that we can call from a new CI job.

My thinking is:

1. Properly testing in WASM will always need some kind of host setup, so it is impossible to do that nicely within the test itself.
2. If we already need a separate CI job and host setup anyway, we might as well do the groundwork for adding more WASM tests.

Thoughts?

Reply:
> Ideally, we only test the public API of such a crate. Unless I am missing something, the two tests you linked to do use only the public API (via the StreamMuxer interface) so I think that should be fine?

You are correct, the connectivity tests can be moved to another crate.

> In regards to the actual tests, I'd be okay with the following:

Ok I will try to do this, however I have some other priorities for now. But I will try to allocate some time for this.

> @thomaseizinger And for making `Transport` `Send`, I meant this: https://github.com/oblique/rust-libp2p/commit/bb55ccab82c9410f0f06641a3d16a5a72d0c3d53
>
> If you prefer it, I can add the commit in this PR.

Can you give me an answer for the above?

Reply:
> > @thomaseizinger And for making `Transport` `Send`, I meant this: [oblique@bb55cca](https://github.com/oblique/rust-libp2p/commit/bb55ccab82c9410f0f06641a3d16a5a72d0c3d53)
> > If you prefer it, I can add the commit in this PR.
> 
> Can you give me an answer for the above?

Sorry I missed this. Yes that looks good to me, please include it.

> > In regards to the actual tests, I'd be okay with the following:
> 
> Ok I will try to do this, however I have some other priorities for now. But I will try to allocate some time for this.

Thank you!

Reply:
I am sorry for commenting only now. I am fine with the suggestion in https://github.com/libp2p/rust-libp2p/pull/4015/files/1c2cafdb426a94699c3c8478e232f774f16a715f#r1234425896. I don't have a strong opinion on this matter.

Reply:
Done